text/0000-cargo-embed-dependency-versions.md

```diff
@@ -0,0 +1,238 @@
+
+# Prior art
+[prior-art]: #prior-art
+
+## In Rust
+
+The Rust compiler already [embeds](https://github.com/rust-lang/rust/pull/97550) compiler and LLVM version in the executables built with it.
+
+An out-of-tree implementation of this RFC exists, see [`cargo auditable`](https://github.com/rust-secure-code/cargo-auditable/), and has garnered considerable interest. NixOS and Void Linux build all their Rust packages with it today; it is also used in production at Microsoft. Extracting the embedded data is already supported by [`rust-audit-info`](https://crates.io/crates/rust-audit-info) and [Syft](https://github.com/anchore/syft). Auditing such binaries for known vulnerabilities is already supported by [`cargo audit`](https://crates.io/crates/cargo-audit) and [Trivy](https://github.com/aquasecurity/trivy).
+
+## In other languages
+
+Suppose a vulnerability that is exploitable over the network is discovered in a popular networking library. We will look at several languages and see how each of them handles it.
+
+### C
+
+Linux distrubutions maintain a strict policy of using shared libraries (dynamic linking) and only having a single copy of the library in the system. The library is also tracked in the package manager.
+
+When the vulnerability is disclosed, an update to the single copy of the library in the system is issued by the distribution. To check if a given system is affected, you only need to look at the version of the distribution package. Tools to scan systems for vulnerable package versions are also available, including as a service.
+
+To mitigate the issue you simply need to run the regular security update command provided by your distribution and reboot. The single copy of the library is replaced and the system is secured.
+
+Sadly this does not apply to packages installed from outside the distribution, or software that is not packaged at all.
+
+### Ruby
+
+Ruby code typically has a `Gemfile.lock` which is analogous to `Cargo.lock`. Crucially, Ruby is never compiled into a binary, and this file is present at runtime in the deployed code. So you can tell exactly what versions of which libraries the code you run actually uses.
+
+Thanks to `Gemfile.lock` the code can be automatically checked for vulnerable package versions against databases such as [RubySec](https://rubysec.com/).
+
+Unlike in the Linux distribution scenario, the dependencies of every program have to be updated individually.
+
+### Go
+
+Go statically links all its code, so the C approach of updating a single instance of a shared library is impossible. Every single Go binary contains a copy of the vulnerable code, and all of them have to be rebuilt.
+
+To make this problem tractable, Go embeds detailed information about the dependency tree in each compiled binary. It is similar to `Cargo.lock` but much more detailed, also including the licenses and other metadata about the dependencies.
```

Comment:
This proposal is focused on security but I imagine corporate legal audits would also appreciate license information.  Should we consider that to any degree?

The number of times we had random binaries when doing legal audits...

Reply:
You can get the license information by looking the crate versions up on crates.io, right? The tool which extracts the dependency info from the executable can have an option to lookup the license too.

Reply:
So long as the dependency is from crates.io.

Reply:
For crates.io, yes. Not all packages originate from crates.io.